File: game_demo/renderer.py
"""ModernGL rendering: context, FBOs, shader programs, draw calls."""

import moderngl
import numpy as np
import glm
from pathlib import Path

from geometry import cube_vertices, quad_vertices


def load_shader(name):
    """
    Pure function. Reads a GLSL shader file from the shaders/ directory.

    Args:
        name (str): Shader filename (e.g. "scene.vert").

    Returns:
        str: Shader source code.

    Examples:
        >>> # load_shader("scene.vert")
    """
    return (Path(__file__).parent / "shaders" / name).read_text()


class Renderer:
    """
    Manages ModernGL context, MRT framebuffer, noise ping-pong, and draw calls.

    Not pure: holds GPU state, renders to framebuffers and screen.

    Args:
        width (int): Framebuffer width in pixels.
        height (int): Framebuffer height in pixels.

    Examples:
        >>> # r = Renderer(800, 600)
    """

    def __init__(self, width, height):
        self.width = width
        self.height = height
        self.ctx = moderngl.create_context()
        self.ctx.enable(moderngl.DEPTH_TEST)
        self.ctx.enable(moderngl.CULL_FACE)
        self.frame_count = 0

        self._init_shaders()
        self._init_geometry()
        self._init_fbos()
        self._init_noise()

    def _init_shaders(self):
        self.scene_prog = self.ctx.program(
            vertex_shader=load_shader("scene.vert"),
            fragment_shader=load_shader("scene.frag"),
        )
        self.noise_warp_prog = self.ctx.program(
            vertex_shader=load_shader("noise_warp.vert"),
            fragment_shader=load_shader("noise_warp.frag"),
        )
        self.display_prog = self.ctx.program(
            vertex_shader=load_shader("display.vert"),
            fragment_shader=load_shader("display.frag"),
        )

    def _init_geometry(self):
        cube_vbo = self.ctx.buffer(cube_vertices().tobytes())
        self.cube_vao = self.ctx.vertex_array(
            self.scene_prog,
            [(cube_vbo, "3f 3f 3f", "in_position", "in_normal", "in_color")],
        )

        quad_data = quad_vertices().tobytes()
        quad_vbo_nw = self.ctx.buffer(quad_data)
        self.noise_quad_vao = self.ctx.vertex_array(
            self.noise_warp_prog,
            [(quad_vbo_nw, "2f 2f", "in_position", "in_texcoord")],
        )
        quad_vbo_disp = self.ctx.buffer(quad_data)
        self.display_quad_vao = self.ctx.vertex_array(
            self.display_prog,
            [(quad_vbo_disp, "2f 2f", "in_position", "in_texcoord")],
        )

    def _init_fbos(self):
        W, H = self.width, self.height

        # Scene MRT: color (RGBA8) + motion vectors (RG32F) + depth
        self.color_tex = self.ctx.texture((W, H), 4)
        self.motion_tex = self.ctx.texture((W, H), 2, dtype="f4")
        self.motion_tex.filter = (moderngl.NEAREST, moderngl.NEAREST)
        self.depth_tex = self.ctx.depth_texture((W, H))
        self.scene_fbo = self.ctx.framebuffer(
            color_attachments=[self.color_tex, self.motion_tex],
            depth_attachment=self.depth_tex,
        )

        # Noise ping-pong (RGBA32F)
        self.noise_tex = [
            self.ctx.texture((W, H), 4, dtype="f4"),
            self.ctx.texture((W, H), 4, dtype="f4"),
        ]
        for t in self.noise_tex:
            t.filter = (moderngl.NEAREST, moderngl.NEAREST)
        self.noise_fbo = [
            self.ctx.framebuffer(color_attachments=[self.noise_tex[0]]),
            self.ctx.framebuffer(color_attachments=[self.noise_tex[1]]),
        ]
        self.noise_idx = 0  # index of the WRITE target

    def _init_noise(self):
        """Upload initial Gaussian noise to both ping-pong textures."""
        noise = np.random.randn(self.height, self.width, 4).astype(np.float32)
        data = noise.tobytes()
        self.noise_tex[0].write(data)
        self.noise_tex[1].write(data)

    def render_scene(self, model, view_proj, prev_model, prev_view_proj):
        """
        Render the cube to the MRT framebuffer (color + motion vectors).

        Not pure: writes to scene_fbo GPU textures.

        Args:
            model (glm.mat4): Current model matrix.
            view_proj (glm.mat4): Current view-projection matrix.
            prev_model (glm.mat4): Previous frame's model matrix.
            prev_view_proj (glm.mat4): Previous frame's view-projection matrix.
        """
        self.scene_fbo.use()
        self.ctx.clear(0.0, 0.0, 0.0, 1.0)

        # PyGLM bytes() is row-major; OpenGL expects column-major → transpose
        self.scene_prog["u_model"].write(bytes(glm.transpose(model)))
        self.scene_prog["u_view_proj"].write(bytes(glm.transpose(view_proj)))
        self.scene_prog["u_prev_model"].write(bytes(glm.transpose(prev_model)))
        self.scene_prog["u_prev_view_proj"].write(bytes(glm.transpose(prev_view_proj)))

        self.cube_vao.render()

    def warp_noise(self):
        """
        Noise warp pass: nearest-neighbor backward warp from motion vectors.

        Reads motion_tex + prev noise, writes new noise via ping-pong.
        Discrete nearest-neighbor lookup — no bilinear interpolation.
        OOB source pixels get fresh PCG Gaussian noise.

        Not pure: reads/writes GPU textures.
        """
        src_idx = 1 - self.noise_idx
        dst_idx = self.noise_idx

        self.noise_fbo[dst_idx].use()
        self.ctx.disable(moderngl.DEPTH_TEST)
        self.ctx.disable(moderngl.CULL_FACE)

        self.motion_tex.use(location=0)
        self.noise_tex[src_idx].use(location=1)

        self.noise_warp_prog["motion_tex"].value = 0
        self.noise_warp_prog["prev_noise_tex"].value = 1
        self.noise_warp_prog["u_resolution"].value = (
            float(self.width),
            float(self.height),
        )
        self.noise_warp_prog["u_frame"].value = self.frame_count

        self.noise_quad_vao.render()

        self.ctx.enable(moderngl.DEPTH_TEST)
        self.ctx.enable(moderngl.CULL_FACE)

        self.noise_idx = 1 - self.noise_idx

    def display(self, mode=0):
        """
        Render final output to the screen.

        Not pure: writes to default framebuffer.

        Args:
            mode (int): 0=noise, 1=color, 2=motion vectors, 3=side-by-side.
        """
        self.ctx.screen.use()
        self.ctx.clear(0.1, 0.1, 0.1, 1.0)
        self.ctx.disable(moderngl.DEPTH_TEST)
        self.ctx.disable(moderngl.CULL_FACE)

        # The noise texture we just wrote to (after ping-pong swap)
        read_idx = 1 - self.noise_idx
        self.noise_tex[read_idx].use(location=0)
        self.color_tex.use(location=1)
        self.motion_tex.use(location=2)

        self.display_prog["noise_tex"].value = 0
        self.display_prog["color_tex"].value = 1
        self.display_prog["motion_tex"].value = 2
        self.display_prog["u_display_mode"].value = mode

        self.display_quad_vao.render()

        self.ctx.enable(moderngl.DEPTH_TEST)
        self.ctx.enable(moderngl.CULL_FACE)

        self.frame_count += 1

    def read_motion(self):
        """
        Read motion vector texture back to CPU.

        Not pure: reads from GPU.

        Returns:
            np.ndarray: Shape [H, W, 2] float32 in (mv_x, mv_y) UV space,
                row 0 = bottom of screen (OpenGL order).

        Examples:
            >>> # mv = renderer.read_motion()
        """
        data = self.motion_tex.read()
        return np.frombuffer(data, dtype=np.float32).reshape(
            self.height, self.width, 2
        )

    def upload_noise(self, noise):
        """
        Upload a CPU noise array to the current write noise texture.

        Not pure: writes GPU texture, swaps ping-pong index.

        Args:
            noise (np.ndarray): [H, W, 4] float32, row 0 = bottom (OpenGL order).

        Examples:
            >>> # renderer.upload_noise(np.random.randn(H, W, 4).astype(np.float32))
        """
        dst_idx = self.noise_idx
        self.noise_tex[dst_idx].write(noise.astype(np.float32).tobytes())
        self.noise_idx = 1 - self.noise_idx
        self.frame_count += 1

    def read_noise(self):
        """
        Read the current noise texture back to CPU.

        Not pure: reads from GPU.

        Returns:
            np.ndarray: Shape [H, W, 4] float32.

        Examples:
            >>> # noise = renderer.read_noise()
        """
        read_idx = 1 - self.noise_idx
        data = self.noise_tex[read_idx].read()
        return np.frombuffer(data, dtype=np.float32).reshape(
            self.height, self.width, 4
        )

File: web_demo/renderer.js
/**
 * WebGL 2 rendering: context, FBOs, shader programs, draw calls.
 * Port of game_demo/renderer.py.
 */

import { cubeVertices, floorVertices, quadVertices } from './geometry.js';
import { sceneVert, sceneFrag, displayVert, displayFrag } from './shaders.js';

const { mat4 } = window.glMatrix;

/**
 * Compile a shader from source. Throws on error.
 * @param {WebGL2RenderingContext} gl
 * @param {number} type - gl.VERTEX_SHADER or gl.FRAGMENT_SHADER
 * @param {string} source
 * @returns {WebGLShader}
 */
function compileShader(gl, type, source) {
    const shader = gl.createShader(type);
    gl.shaderSource(shader, source);
    gl.compileShader(shader);
    if (!gl.getShaderParameter(shader, gl.COMPILE_STATUS)) {
        const info = gl.getShaderInfoLog(shader);
        gl.deleteShader(shader);
        throw new Error('Shader compile error: ' + info);
    }
    return shader;
}

/**
 * Link a program from vertex + fragment shaders. Throws on error.
 * @param {WebGL2RenderingContext} gl
 * @param {WebGLShader} vert
 * @param {WebGLShader} frag
 * @returns {WebGLProgram}
 */
function linkProgram(gl, vert, frag) {
    const prog = gl.createProgram();
    gl.attachShader(prog, vert);
    gl.attachShader(prog, frag);
    gl.linkProgram(prog);
    if (!gl.getProgramParameter(prog, gl.LINK_STATUS)) {
        const info = gl.getProgramInfoLog(prog);
        gl.deleteProgram(prog);
        throw new Error('Program link error: ' + info);
    }
    return prog;
}

/**
 * Create a shader program from vertex + fragment source strings.
 * @param {WebGL2RenderingContext} gl
 * @param {string} vertSrc
 * @param {string} fragSrc
 * @returns {WebGLProgram}
 */
function createProgram(gl, vertSrc, fragSrc) {
    const v = compileShader(gl, gl.VERTEX_SHADER, vertSrc);
    const f = compileShader(gl, gl.FRAGMENT_SHADER, fragSrc);
    return linkProgram(gl, v, f);
}

export class Renderer {
    /**
     * @param {HTMLCanvasElement} canvas
     * @param {number} width
     * @param {number} height
     */
    constructor(canvas, width, height) {
        this.width = width;
        this.height = height;
        this.frameCount = 0;

        const gl = canvas.getContext('webgl2', { antialias: false });
        if (!gl) throw new Error('WebGL 2 not available');
        this.gl = gl;

        // Required for float FBOs
        const ext = gl.getExtension('EXT_color_buffer_float');
        if (!ext) throw new Error('EXT_color_buffer_float not available');

        gl.enable(gl.DEPTH_TEST);
        gl.enable(gl.CULL_FACE);

        this._initShaders();
        this._initGeometry();
        this._initFBOs();

        // Pre-allocate readback buffer for motion vectors
        this._motionReadback = new Float32Array(width * height * 4);
    }

    _initShaders() {
        const { gl } = this;
        this.sceneProg = createProgram(gl, sceneVert, sceneFrag);
        this.displayProg = createProgram(gl, displayVert, displayFrag);

        // Cache uniform locations — scene
        this.sceneUniforms = {
            u_model:           gl.getUniformLocation(this.sceneProg, 'u_model'),
            u_view_proj:       gl.getUniformLocation(this.sceneProg, 'u_view_proj'),
            u_prev_model:      gl.getUniformLocation(this.sceneProg, 'u_prev_model'),
            u_prev_view_proj:  gl.getUniformLocation(this.sceneProg, 'u_prev_view_proj'),
        };

        // Cache uniform locations — display
        this.displayUniforms = {
            noise_tex:      gl.getUniformLocation(this.displayProg, 'noise_tex'),
            color_tex:      gl.getUniformLocation(this.displayProg, 'color_tex'),
            motion_tex:     gl.getUniformLocation(this.displayProg, 'motion_tex'),
            u_display_mode: gl.getUniformLocation(this.displayProg, 'u_display_mode'),
        };
    }

    _initGeometry() {
        const { gl } = this;

        // Cube VAO
        this.cubeVAO = gl.createVertexArray();
        gl.bindVertexArray(this.cubeVAO);
        const cubeVBO = gl.createBuffer();
        gl.bindBuffer(gl.ARRAY_BUFFER, cubeVBO);
        gl.bufferData(gl.ARRAY_BUFFER, cubeVertices(), gl.STATIC_DRAW);
        this._setupSceneAttribs();
        gl.bindVertexArray(null);

        // Floor VAO
        this.floorVAO = gl.createVertexArray();
        gl.bindVertexArray(this.floorVAO);
        const floorVBO = gl.createBuffer();
        gl.bindBuffer(gl.ARRAY_BUFFER, floorVBO);
        gl.bufferData(gl.ARRAY_BUFFER, floorVertices(), gl.STATIC_DRAW);
        this._setupSceneAttribs();
        gl.bindVertexArray(null);

        // Display quad VAO
        this.displayQuadVAO = gl.createVertexArray();
        gl.bindVertexArray(this.displayQuadVAO);
        const quadVBO = gl.createBuffer();
        gl.bindBuffer(gl.ARRAY_BUFFER, quadVBO);
        gl.bufferData(gl.ARRAY_BUFFER, quadVertices(), gl.STATIC_DRAW);
        this._setupQuadAttribs(this.displayProg);
        gl.bindVertexArray(null);
    }

    _setupSceneAttribs() {
        const { gl } = this;
        const stride = 9 * 4; // 9 floats * 4 bytes
        const posLoc = gl.getAttribLocation(this.sceneProg, 'in_position');
        const normLoc = gl.getAttribLocation(this.sceneProg, 'in_normal');
        const colLoc = gl.getAttribLocation(this.sceneProg, 'in_color');
        gl.enableVertexAttribArray(posLoc);
        gl.vertexAttribPointer(posLoc, 3, gl.FLOAT, false, stride, 0);
        gl.enableVertexAttribArray(normLoc);
        gl.vertexAttribPointer(normLoc, 3, gl.FLOAT, false, stride, 12);
        gl.enableVertexAttribArray(colLoc);
        gl.vertexAttribPointer(colLoc, 3, gl.FLOAT, false, stride, 24);
    }

    _setupQuadAttribs(prog) {
        const { gl } = this;
        const stride = 4 * 4;
        const posLoc = gl.getAttribLocation(prog, 'in_position');
        const tcLoc = gl.getAttribLocation(prog, 'in_texcoord');
        gl.enableVertexAttribArray(posLoc);
        gl.vertexAttribPointer(posLoc, 2, gl.FLOAT, false, stride, 0);
        gl.enableVertexAttribArray(tcLoc);
        gl.vertexAttribPointer(tcLoc, 2, gl.FLOAT, false, stride, 8);
    }

    _initFBOs() {
        const { gl, width: W, height: H } = this;

        // --- Scene MRT: color (RGBA8) + motion (RGBA32F) + depth ---
        this.colorTex = gl.createTexture();
        gl.bindTexture(gl.TEXTURE_2D, this.colorTex);
        gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA8, W, H, 0, gl.RGBA, gl.UNSIGNED_BYTE, null);
        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.NEAREST);
        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.NEAREST);

        // Motion vectors: use RGBA32F (readPixels requires 4 components)
        this.motionTex = gl.createTexture();
        gl.bindTexture(gl.TEXTURE_2D, this.motionTex);
        gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA32F, W, H, 0, gl.RGBA, gl.FLOAT, null);
        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.NEAREST);
        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.NEAREST);

        this.depthRB = gl.createRenderbuffer();
        gl.bindRenderbuffer(gl.RENDERBUFFER, this.depthRB);
        gl.renderbufferStorage(gl.RENDERBUFFER, gl.DEPTH_COMPONENT24, W, H);

        this.sceneFBO = gl.createFramebuffer();
        gl.bindFramebuffer(gl.FRAMEBUFFER, this.sceneFBO);
        gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT0, gl.TEXTURE_2D, this.colorTex, 0);
        gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.COLOR_ATTACHMENT1, gl.TEXTURE_2D, this.motionTex, 0);
        gl.framebufferRenderbuffer(gl.FRAMEBUFFER, gl.DEPTH_ATTACHMENT, gl.RENDERBUFFER, this.depthRB);
        gl.drawBuffers([gl.COLOR_ATTACHMENT0, gl.COLOR_ATTACHMENT1]);

        const status = gl.checkFramebufferStatus(gl.FRAMEBUFFER);
        if (status !== gl.FRAMEBUFFER_COMPLETE) {
            throw new Error('Scene FBO incomplete: 0x' + status.toString(16));
        }

        // --- Noise texture for display (uploaded from JS each frame) ---
        this.noiseTex = gl.createTexture();
        gl.bindTexture(gl.TEXTURE_2D, this.noiseTex);
        gl.texImage2D(gl.TEXTURE_2D, 0, gl.RGBA32F, W, H, 0, gl.RGBA, gl.FLOAT, null);
        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.NEAREST);
        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.NEAREST);

        gl.bindFramebuffer(gl.FRAMEBUFFER, null);
    }

    /**
     * Render cube + floor to the MRT framebuffer.
     * @param {Float32Array} model - 4x4 column-major
     * @param {Float32Array} viewProj - 4x4 column-major
     * @param {Float32Array} prevModel - 4x4 column-major
     * @param {Float32Array} prevViewProj - 4x4 column-major
     */
    renderScene(model, viewProj, prevModel, prevViewProj) {
        const { gl } = this;
        gl.bindFramebuffer(gl.FRAMEBUFFER, this.sceneFBO);
        gl.viewport(0, 0, this.width, this.height);
        gl.clearColor(0, 0, 0, 1);
        gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);
        gl.enable(gl.DEPTH_TEST);
        gl.enable(gl.CULL_FACE);
        gl.useProgram(this.sceneProg);

        const identity = mat4.create();  // gl-matrix identity

        // Floor (static — identity model, zero motion)
        gl.uniformMatrix4fv(this.sceneUniforms.u_model, false, identity);
        gl.uniformMatrix4fv(this.sceneUniforms.u_view_proj, false, viewProj);
        gl.uniformMatrix4fv(this.sceneUniforms.u_prev_model, false, identity);
        gl.uniformMatrix4fv(this.sceneUniforms.u_prev_view_proj, false, prevViewProj);
        gl.bindVertexArray(this.floorVAO);
        gl.drawArrays(gl.TRIANGLES, 0, 6);

        // Cube
        gl.uniformMatrix4fv(this.sceneUniforms.u_model, false, model);
        gl.uniformMatrix4fv(this.sceneUniforms.u_prev_model, false, prevModel);
        gl.bindVertexArray(this.cubeVAO);
        gl.drawArrays(gl.TRIANGLES, 0, 36);

        gl.bindVertexArray(null);
    }

    /**
     * Read motion vectors from the scene FBO.
     * @returns {Float32Array} [H * W * 2] in (mv_x, mv_y) order, row 0 = bottom
     */
    readMotion() {
        const { gl, width: W, height: H } = this;
        gl.bindFramebuffer(gl.FRAMEBUFFER, this.sceneFBO);
        // Read from motion attachment (COLOR_ATTACHMENT1)
        gl.readBuffer(gl.COLOR_ATTACHMENT1);
        gl.readPixels(0, 0, W, H, gl.RGBA, gl.FLOAT, this._motionReadback);

        // Extract RG channels from RGBA → [H*W*2]
        const out = new Float32Array(W * H * 2);
        for (let i = 0; i < W * H; i++) {
            out[i * 2]     = this._motionReadback[i * 4];      // mv_x
            out[i * 2 + 1] = this._motionReadback[i * 4 + 1];  // mv_y
        }
        return out;
    }

    /**
     * Upload noise data to the noise texture for display.
     * @param {Float32Array} data - [H * W * 4] RGBA float, row 0 = bottom (OpenGL order)
     */
    uploadNoise(data) {
        const { gl, width: W, height: H } = this;
        gl.bindTexture(gl.TEXTURE_2D, this.noiseTex);
        gl.texSubImage2D(gl.TEXTURE_2D, 0, 0, 0, W, H, gl.RGBA, gl.FLOAT, data);
        this.frameCount++;
    }

    /**
     * Render final output to the canvas.
     * @param {number} mode - 0=noise, 1=color, 2=motion, 3=side-by-side, 4=raw
     */
    display(mode = 0) {
        const { gl } = this;
        gl.bindFramebuffer(gl.FRAMEBUFFER, null);
        gl.viewport(0, 0, this.width, this.height);
        gl.clearColor(0.1, 0.1, 0.1, 1);
        gl.clear(gl.COLOR_BUFFER_BIT);
        gl.disable(gl.DEPTH_TEST);
        gl.disable(gl.CULL_FACE);

        gl.useProgram(this.displayProg);

        // Bind textures
        gl.activeTexture(gl.TEXTURE0);
        gl.bindTexture(gl.TEXTURE_2D, this.noiseTex);
        gl.activeTexture(gl.TEXTURE1);
        gl.bindTexture(gl.TEXTURE_2D, this.colorTex);
        gl.activeTexture(gl.TEXTURE2);
        gl.bindTexture(gl.TEXTURE_2D, this.motionTex);

        gl.uniform1i(this.displayUniforms.noise_tex, 0);
        gl.uniform1i(this.displayUniforms.color_tex, 1);
        gl.uniform1i(this.displayUniforms.motion_tex, 2);
        gl.uniform1i(this.displayUniforms.u_display_mode, mode);

        gl.bindVertexArray(this.displayQuadVAO);
        gl.drawArrays(gl.TRIANGLES, 0, 6);

        gl.bindVertexArray(null);
        gl.enable(gl.DEPTH_TEST);
        gl.enable(gl.CULL_FACE);
    }
}
